chaos-days/blog/2023-12-19-Dynamic-Scaling-with-Dataloss/index.md

```diff
@@ -0,0 +1,143 @@
+        "PartitionId": 5,
+        "Priority": 2
+      },
+      ...
+    ]
+  }
+}
+```
+
+The coordinator is a member of partition 5 but there are two remaining members of partition 5 that should allow broker 5 to join.
+![](p5-roles.png)
+
+After restoring the coordinator again with `zbchaos dataloss recover --nodeId 0`, joining eventually completes and the scaling operation finishes successfully.
+
+![](scaleup-complete.png)
+
+```json
+{
+  "Version": 19,
+  "LastChange": {
+    "Id": 18,
+    "Status": "COMPLETED",
+    "StartedAt": "2023-12-18T17:05:55.849442157Z",
+    "CompletedAt": "2023-12-18T17:17:32.913050015Z"
+  },
+  "PendingChange": null
+}
+```
+
+Overall, we achieve our goal that the scaling operation eventually completes with the expected result.
+The coordinator recovers after dataloss and eventually receives the cluster topology from the remaining brokers.
+
+However, it was unexpected that the scaling did not make progress while the coordinator was down.
+
+## Single point of failure during scaling
+
+The issue that scaling did not make progress while the coordinator was reproducible.
+Eventually, we diagnosed it as the following edge case:
+
+When scaling up and and adding a new member to the replication group of a partition, the raft partition goes through joint consensus.
+The details of this process are described in the [raft paper](https://raft.github.io/raft.pdf), but here is a very short summary:
+Joint consensus is similar to a 2-phase commit, where the leader of the partition first introduces a new _joint consensus_ configuration that requires quorum from both the old and new set of members.
+After committing the joint consensus configuration, the leader leaves joint consensus by "forgetting" the old member set and only using the new member set.
+Only after this second configuration is committed, joining of the new member is complete.
+
+In our example, the new set of members has size 4, one of which is the coordinator and one is the newly joining member.
+With 4 members, the quorum is 3, meaning that the partition can only elect a leader and process if at least 3 members are available.
+In our experiment, we made the coordinator unavailable, so we were already down to 3 members.
+Additionally, the newly joining member did not start yet because it was waiting for a successful join response from the leader.
+The newly joining member never received such a response because we took down the coordinator which was the leader of the partition.
+This resulted in only 2 out of 4 members being available, which is not enough to elect a leader.
+
+We want to improve this behavior in the future by allowing responding earlier to join requests.
+Currently, a response is only sent after leaving the joint consensus phase.
+To reduce the likelihood that the newly joining member is unavailable because it waits for a response, we can send the join response earlier, right after entering the joint consensus phase.
```

Comment:
❓ but this still can end up in the same situation right? Is there a way to completely avoid this?

Is this problem only because of the datalass so problematic or would this also happen on normal restarts during scaling? 🤔

Reply:
Yeah, it's not completely avoided, just the chance is reduced. To avoid it completely is too tricky for now, but it's also not critical since we can assume that eventually the missing broker will come back.
The problem is not dataloss, it's just unavailability.